flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/ArrayToArrayCastRule.java

```diff
@@ -109,7 +110,7 @@ protected String generateCodeBlockInternal(
 
     private static String arrayElementType(LogicalType t) {
         if (t.isNullable()) {
-            return "Object";
+            return CodeGenUtils.boxedTypeTermForType(t);
         }
         switch (t.getTypeRoot()) {
             case BOOLEAN:
```

Comment:
hi, I think it would be better to use CodeGenUtils.primitiveTypeTermForType instead, what do you think? 
Maybe we need to check isPrimitiveNullable first when t is nullable, if it is true, we should use primitiveType instead of boxedType.

Reply:
I pushed a change, also adding another test to make sure this works. Does it looks good now?